### PR TITLE
Allow forward references in computed properties of type declarations

### LINKED ISSUE #23829: Unnecessary reference checks with interfaces
cc @weswigham 

**TypeScript Version:**  master

**Search Terms:** 

**Code**

```ts
interface C {
  [C.a](): void;
  [C.b]: void;
}
class C {
  static readonly a = Symbol();
  static readonly b = Symbol();
}
```

**Expected behavior:**

pass

**Actual behavior:**

```
$ node built/local/tsc.js index.ts -t esnext
index.ts:3:4 - error TS2449: Class 'C' used before its declaration.

3   [C.b]: void;
     ~
```

**Playground Link:**  

**Related Issues:**

### PR DESCRIPTION
**Pull request by Andarist**
fixes #23829

this is sort-of a followup to https://github.com/microsoft/TypeScript/pull/37646, this PR didn't cover all the cases and it didn't work out properly with `strictNullChecks: true` (or it has regressed on that front since then)

### CODE REVIEW

**Comment by jakebailey** [comment]
@typescript-bot user test this inline
@typescript-bot perf test this faster
@typescript-bot test top100

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23472,6 +23468,10 @@ namespace ts {
                 n => n.kind === SyntaxKind.TypeQuery ? true : n.kind === SyntaxKind.Identifier || n.kind === SyntaxKind.QualifiedName ? false : "quit");
         }
 
+        function isInTypeDeclaration(node: Node): boolean {
+            return !!findAncestor(node, n => isInterfaceDeclaration(n) || isTypeLiteralNode(n) || !!(n.flags & NodeFlags.Ambient));
```
[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts:23472]: if the Ambient flag would be propagated to all children then the `findAncestor` call would be redundant, right? 🤔 or did I misunderstand what you meant?

Either way, regardless of that - it makes sense to reorder those checks for perf reasons

[the remark this answers is not in the document]
Comment (Andarist) [reply src/compiler/checker.ts]: Oh, when reading in the squeezed GitHub UI I've misread my original code and I thought that the ambient check was already outside of the `findAncestor` call but it was actually in it. Disregard the previous comment - I've applied the suggested change

**Comment by sandersn** [comment]
New changes look good, although I still want to know why the call to isInTypeDeclaration is needed inside checkIdentifier.

**Comment by Andarist** [comment]
@sandersn weird, I definitely posted the answer to this one yesterday but it's gone 🤷‍♂️ reposted it now, please take a look.

**Review by sandersn — APPROVED** [review]

**Pull request merged**